renderer/Profiling.cpp

```diff
@@ -0,0 +1,181 @@
+/*****************************************************************************
+The Dark Mod GPL Source Code
+
+This file is part of the The Dark Mod Source Code, originally based
+on the Doom 3 GPL Source Code as published in 2011.
+
+The Dark Mod Source Code is free software: you can redistribute it
+and/or modify it under the terms of the GNU General Public License as
+published by the Free Software Foundation, either version 3 of the License,
+or (at your option) any later version. For details, see LICENSE.TXT.
+
+Project: The Dark Mod (http://www.thedarkmod.com/)
+
+******************************************************************************/
+#include "precompiled.h"
+#pragma hdrstop
+
+#include "Profiling.h"
+
+idCVar r_useDebugGroups( "r_useDebugGroups", "0", CVAR_RENDERER | CVAR_BOOL, "Emit GL debug groups during rendering. Useful for frame debugging and analysis with e.g. nSight, which will group render calls accordingly." );
+idCVar r_glProfiling( "r_glProfiling", "0", CVAR_RENDERER | CVAR_BOOL, "Collect profiling information about GPU and CPU time spent in the rendering backend." );
+
+/**
+ * Tracks CPU and GPU time spent in profiled sections.
+ * Sections are structured in hierarchical fashion, you can push and pop sections onto a stack.
+ * Each section has a name; sections with the same name at the same place in the hierarchy will
+ * be accumulated to a single total time in the end.
+ */
+class GlProfiler {
+public:
+	GlProfiler() : profilingActive(false), frameMarker(0) {}
+
+	void BeginFrame() {
+		sectionStack.clear();
+		frame[ frameMarker ] = section();
+		frame[ frameMarker ].name = "Frame";
+		AddProfilingQuery( &frame[ frameMarker ] );
+		sectionStack.push_back( &frame[ frameMarker ] );
+		profilingActive = true;
+	}
+
+	void EndFrame() {
+		LeaveSection();
+		assert( sectionStack.empty() );
+		frameMarker = ( frameMarker + 1 ) % 2;
+		profilingActive = false;
+	}
+
+	void EnterSection(const idStr& name) {
+		if( !profilingActive )
+			return;
+		assert( !sectionStack.empty() );
+		section *s = FindOrInsertSection( name );
+		AddProfilingQuery( s );
+		sectionStack.push_back( s );
+	}
+
+	void LeaveSection() {
+		if( !profilingActive )
+			return;
+		assert( !sectionStack.empty() );
+		CompleteProfilingQuery( sectionStack.back() );
+		sectionStack.pop_back();
+	}
+
+	typedef struct query {
+		uint64_t cpuStartTime;
+		uint64_t cpuStopTime;
+		GLuint glQueries[ 2 ];
+	} query;
+
+	typedef struct section {
+		idStr name;
+		int count;
+		double totalCpuTimeMillis;
+		double totalGpuTimeMillis;
+		std::vector<section> children;
+		std::vector<query> queries;
+	} section;
+
+	section * CollectTimings() {
+		AccumulateTotalTimes( frame[ frameMarker ] );
```

Comment:
Do I understand right that here you request timings of the **current** frame?
If yes, than how can it display CPU timings correctly?
What if this function is called when profiler is active?

It seems that the most non-intrusive policy is to display results of the previous or one-before-previous frame, while storing more than two consecutive frames.
Fetching results of the just finished frame is not a very good idea, because if the rendering of that frame is not yet finished, then quering gl timers would block CPU (If I understand correctly how it works).

Reply:
At least add here assert that profiling is not active.
This will ensure that the frame you are processing here is already finished (from CPU prespective).